### PR TITLE
Add discriminated union support (v2)

### LINKED ISSUE #4675: V2: Descriminated Unions
Descriminated unions are getting lots of super powers in V2, probably best example right now is the tests in pydantic-core, see https://github.com/pydantic/pydantic-core/blob/main/tests/validators/test_tagged_union.py.

In particular, descriminated union desciminators can now be a path or a function.

How should we define these? I guess it can all be done with `Field` and `Annotated`?

### PR DESCRIPTION
**Pull request by dmontagu**
**Note to self:** Once PR #5125 is merged, I'll want to merge in the changes from the `v2-discriminated-union-generics` branch.

There are still some remaining challenges, but I think this is mostly good:
* Handling models with `__root__` (I think this is getting removed in v2 though)
* Handling aliases properly (will do that in the future once more things with alias are handled properly)

fix #4675

### CODE REVIEW

File: tests/test_json_schema.py
Comment (dmontagu) [inline tests/test_json_schema.py:2685]: @Julian Example 1:

File: tests/test_json_schema.py
Comment (dmontagu) [inline tests/test_json_schema.py:2812]: @Julian Example 2 (more complex):

File: tests/test_json_schema.py
Comment (dmontagu) [inline tests/test_json_schema.py:2913]: @Julian Example 3 (even more complex):

**Comment by dmontagu** [comment]
@Julian I called out a few cases of more schemas in the tests making use of the OpenAPI discriminator. (Will probably be better to view the final file than the diff in the files view). I recognize this is OpenAPI-specific, but I would be interested to get your take on:

1. Would you consider this to be valid JSON schema? (And appropriate for the relevant types.)
    * If not, is there anything I can do to help it be more compliant? (E.g., something something vocabularies/extensions/???, or maybe adding a flag for whether to include the discriminator stuff in the generated schema, i.e., whether to include openapi extensions.)
2. Does this seem to be valid for OpenAPI (not sure if that's your wheelhouse or not)?

I did run some examples through some online JSON schema validators and it seemed to respect the discriminator stuff properly, but I don't necessarily view that as authoritative 😄.

File: tests/test_json_schema.py
Comment (Julian) [reply tests/test_json_schema.py:2685]: I'm staring strictly at the JSON Schemas here not so much the implementation code, so if anything I say is ugly to actually implement that's perfectly believable! And indeed I'm no expert (yet?) on OpenAPI even so hope I don't say anything too ridiculous.

File: tests/test_json_schema.py
Comment (Julian) [inline tests/test_json_schema.py:2689]: I think a human authored schema is a lot more likely to spec out `pet_type` a level up, to ensure it's always uniformly typed. I.e. I'd probably hand-write something like:

```
  '$defs': {
            'Cat': {'const': 'cat'},
            'Dog': {'const': 'dog'},
            'Lizard': {'enum': ['reptile', 'lizard']},
  },
  "properties": {
    "pet": {
      "required": ["pet_type"],
      "properties": {
        "pet_type": {
          "type": "string",
          "oneOf": {
            {"$ref": "#/$defs/Cat"},
            {"$ref": "#/$defs/Dog"},
            {"$ref": "#/$defs/Lizard"},
          }
       }
  }
  
```

(As I say whether that's a ton more difficult implementation-wise I haven't considered)

But what you have indeed looks valid to me.

Maybe one implementation note is you may consider in your test suite having a guard assertion that the schema you generated is valid (shameless plug obviously, but that's calling [`Draft202012Validator.check_schema(whatyouhavehere)`](https://python-jsonschema.readthedocs.io/en/stable/api/jsonschema/protocols/#jsonschema.protocols.Validator.check_schema) in my implementation).

File: tests/test_json_schema.py
Comment (Julian) [inline tests/test_json_schema.py:2721]: You have `Lizard` twice here, I guess because of the `lizard`/`reptile` business. This is going to fail then anytime someone has a lizard -- because anytime it matches one of those, it will match the second too (and `oneOf` allows only one match). I'm not sure what the intended output is supposed to be, whether you just want to deduplicate before emitting this `oneOf` or if this indeed is meant to always fail.

File: tests/test_json_schema.py
Comment (Julian) [inline tests/test_json_schema.py:2820]: (For more complex schemas I think it's even more attractive to say these would normally be a "level-up" in the schema so that they're not repeated each time...)

**Review by Julian — COMMENTED** [review]
Left a comment or two which hopefully is helpful to think about, though may be undesirable implementation-wise.

Happy to keep looking at pings of course!

**Comment by dmontagu** [comment]
@samuelcolvin Would you be open to adding `jsonschema` as a testing dependency for the sake of validating generated schemas? I like the idea, if for no other reason than making sure we recognize if/when we make a change that isn't compatible with whatever draft we target (i.e., 2020-12 for now).

File: tests/test_json_schema.py
Comment (dmontagu) [reply tests/test_json_schema.py:2721]: Good catch, I think I figured out how to resolve this

**Comment by samuelcolvin** [comment]
> @samuelcolvin Would you be open to adding `jsonschema` as a testing dependency for the sake of validating generated schemas? I like the idea, if for no other reason than making sure we recognize if/when we make a change that isn't compatible with whatever draft we target (i.e., 2020-12 for now).

yup, fine by me.

**Comment by dmontagu** [comment]
This should be ready for review now, there's nothing else I intend to do before merging (unless I get feedback to the contrary).

File: pydantic/_internal/_generate_schema.py
Comment (adriangb) [inline pydantic/_internal/_generate_schema.py:874]: Do we also need to handle the case of a nullable schema here? Any other cases we can "collapse" into a union? I'm not sure if it happens earlier but I wonder how we handle:

```python
Result =T | Exception
MaybeIntResult = Result[int] | None
```

File: pydantic/json_schema.py
Comment (adriangb) [inline pydantic/json_schema.py:22]: Can we make these private?

**Review by adriangb — APPROVED** [review]

File: pydantic/_internal/_generate_schema.py
Comment (dmontagu) [reply pydantic/_internal/_generate_schema.py:797]: Yes, I think this is described in pydantic_core somewhere (I wanted documentation because it confused the crap out of me when I first saw it), but I'll add a note here about what it means. The idea for discriminators in general:
* `str`: name of property to grab to obtain discriminator value
* `int`: index to grab (for arrays) to obtain discriminator value
* `list[str | int]`: obtain the discriminator value by iteratively/recursively applying the items of the list to the value obtained following the instructions in the first two bullets above
* `list[list[str | int]]`: follow the procedure described in the previous bullet for each item in the outer list until you get a valid discriminator value

So, if you need to check multiple fields to grab the discriminator value (as is the case if populate_by_name is True and there is an alias), the way to do this is to use the list-of-lists approach (representing alternative choices) where the inner lists represent accessing the specific property that you want to grab the discriminator from.

Does that make sense?

File: pydantic/_internal/_generate_schema.py
Comment (dmontagu) [reply pydantic/_internal/_generate_schema.py:828]: I am purposely making a copy of the list so I can pop items out below. I reversed it so that the "popping" below produces items in the same order they occurred in the list (since they will pop off the end instead of from the start).

That said, writing this, I realize I need to reverse the items below while extending in the coalescing case.

File: pydantic/_internal/_generate_schema.py
Comment (dmontagu) [reply pydantic/_internal/_generate_schema.py:835]: I'll add a comment explaining what is happening

File: pydantic/_internal/_generate_schema.py
Comment (adriangb) [reply pydantic/_internal/_generate_schema.py:797]: Wow, I mean yeah that makes sense but I'm surprised we support something that complex. Do other libraries (zod, etc.) support this sort of recursive discriminant stuff? I thought most only supported the "give me the name of a property, it better be a scalar value" type thing.

File: pydantic/_internal/_generate_schema.py
Comment (dmontagu) [reply pydantic/_internal/_generate_schema.py:790]: 🤷‍♂️ I think there was a test for this; either way I was just replicating the same logic that was in v1. I agree it doesn't seem there's anything wrong with a one-item union in general.

File: pydantic/_internal/_generate_schema.py
Comment (adriangb) [reply pydantic/_internal/_generate_schema.py:828]: I would still accept a list an make a copy. It may not be a problem here but IMO the issue with this sort of API is that the caller may assume you're being lazy processing the items and thus give you `range(1, 1_000_000_000_000)` but then if you try to put that all into memory kaboom.

File: pydantic/_internal/_generate_schema.py
Comment (adriangb) [reply pydantic/_internal/_generate_schema.py:790]: If there's a test for it then fair to leave it as is. In general I'd prefer to handle cases like this by "doing nothing special" unless there's some really bad implications for the users.

File: pydantic/_internal/_generate_schema.py
Comment (dmontagu) [reply pydantic/_internal/_generate_schema.py:840]: It does a bit more than this, including updating the `aliases` value. I will add a docstring explaining.

File: pydantic/_internal/_generate_schema.py
Comment (dmontagu) [reply pydantic/_internal/_generate_schema.py:874]: good catch. I'll see if I can produce an error in a new test with the current code.

File: pydantic/_internal/_generate_schema.py
Comment (dmontagu) [reply pydantic/_internal/_generate_schema.py:790]: Not going to change this for now since, even though the error messages are the same, you'd be able to tell which error you hit in the traceback by looking at the line numbers. I think we should probably change the error message for these two lines anyway but perhaps that can come later with a copy-pass over all of our error messages.

File: pydantic/_internal/_generate_schema.py
Comment (dmontagu) [reply pydantic/_internal/_generate_schema.py:797]: Most things that work with discriminators definitely work that way, I think Samuel implemented this in pursuit of generality. That said, I think this will allow people to create nice wrappers around certain "crappy" APIs that don't do "standard" things with discriminators.

Also note: the pydantic_core TaggedUnion thing actually even allows arbitrary python functions to be used to obtain the discriminator from a value. I haven't added any support for that now though.

File: pydantic/_internal/_generate_schema.py
Comment (adriangb) [reply pydantic/_internal/_generate_schema.py:828]: Also reversing the list will copy it:

```
>>> ls = [1, 2, 3]
>>> rv = ls[::-1]
>>> rv.pop()
1
>>> ls
[1, 2, 3]
```

Comment (dmontagu) [reply pydantic/_internal/_generate_schema.py:828]: I realized that too, but sadly also discovered this mis-typing issue: https://github.com/pydantic/pydantic-core/issues/427

**Comment by dmontagu** [comment]
please review

**Comment by okke-formsma** [comment]
I've been fighting with pydantic + jsonschema for a bit, and one of the reasons is that pydantic claims to support json-schema, there are some things like discriminated unions in it that are openAPI-specific.

For the unexpecting user of Json Schema and Pydantic, it is rather surprising that non-standard json schema extensions are included without them being labeled as such. 

Should v2 of pydantic not have a "target" or "type" option in the `schema` method that specifies the target to build the schema for? (e.g., 'OpenAPI' or 'json-schema' or maybe even 'json-schema-2020-12' for when they inevitably come up with another non-backward-compatible json schema release)

**Comment by dmontagu** [comment]
@okke-formsma I think you aren't the first person to mention wanting the ability to request specific JSON schema behavior along these lines. Fortunately, we've made a lot of improvements to the modularity of the JSON schema generation as part of v2, and it should be much easier to modify it to suit various needs (whether we end up supporting that directly or leave it to user code).

There are a handful of more-minor JSON schema issues that still need to be resolved before v2 is released, and the integration with FastAPI still needs to happen. I am not exactly sure what the best API would be yet for controlling the JSON schema behavior, but I will note that it should be pretty easy to exclude the discriminated union stuff in user code currently on V2 — if I understand how JSON schema should work correctly, you'd just need to subclass GenerateJsonSchema and replace the current logic for handling tagged unions with (essentially) the same logic used in regular unions. There is a semi-related example of how to override JSON schema generation in v2 in `tests/test_fastapi_json_schema.py`.

If you have opinions about the API you'd like for controlling what kind of (pseudo?-)JSON schema should be produced (e.g., with or without OpenAPI), feel free to create an issue about it. I would be open to including a strictly-JSON-schema generating class as an alternative to what is currently called `GenerateJsonSchema`, and possibly tweaking the method signature for `model_json_schema` to perhaps have some special allowed string values for `schema_generator` for some "official" choices.

All that said, I think for the sake of this PR I don't want to get too caught up with the JSON schema details, since this PR is more focused on achieving parity with the v1 behavior. But definitely happy to put more thought into how we can improve the JSON schema APIs before v2 is released.

**Comment by okke-formsma** [comment]
Thanks for your response. That sounds like a pretty good solution. I'm definitely interested in the ability to subclass the schema generation, but I'll chime in in a more appropriate place.

**Comment by Kludex** [comment]
> Thanks for your response. That sounds like a pretty good solution. I'm definitely interested in the ability to subclass the schema generation, but I'll chime in in a more appropriate place.

I was talking to @dmontagu , and he implemented it on a conversation we had:

```python
from pydantic_core import core_schema

from pydantic.json_schema import GenerateJsonSchema, JsonSchemaValue


class GenerateStrictJsonSchema(GenerateJsonSchema):
    def tagged_union_schema(
        self, schema: core_schema.TaggedUnionSchema
    ) -> JsonSchemaValue:
        schema = schema.copy()
        schema.pop("discriminator")
        schema = dict(
            **schema,
            type="union",
            choices=[x for x in schema["choices"].values() if isinstance(x, dict)]
        )
        return self.union_schema(schema)
```

The above should be enough. 🙏

**Pull request merged**
